### PR TITLE
Support for PKCS#12

### PR DESCRIPTION
**Pull request by aleks-f**
- [x] PKCS12Container
- [x] EC key
- [x] ECDSA digest
- [x] tests

### CODE REVIEW

File: Crypto/src/PKCS12Container.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+void PKCS12Container::load(PKCS12* pPKCS12, const std::string& password)
+{
+	if (pPKCS12)
+	{
+		X509* pCert = nullptr;
+		STACK_OF(X509)* pCA = nullptr;
+		STACK_OF(PKCS12_SAFEBAG)* bags = nullptr;
+		if (PKCS12_parse(pPKCS12, password.c_str(), &_pKey, &pCert, &pCA))
+		{
+			if (pCert && _pKey)
```
Comment (bachp) [inline Crypto/src/PKCS12Container.cpp]: I think it is possible to hava a PKCS12 container without a key. (See: `trusted.p12`)

I created some test files: [ca.zip](https://github.com/pocoproject/poco/files/1291993/ca.zip)


`server.p12` containing:
- server.key - private key
- server.pem - certificate
- ca-chain.pem - certificate chain with root and intermediate
created: `openssl pkcs12 -export -out server.p12 -inkey intermediate/private/server.key.pem -in intermediate/certs/server.cert.pem -certfile intermediate/certs/ca-chain.cert.pem -name vally`
password: `crypto`
info: `openssl pkcs12 -info -in server.p12`


`trusted.p12` containing:
- list of certificates
created: `openssl pkcs12 -export -out trusted.p12 -in intermediate/certs/ca-chain.cert.pem -certfile trusted-certs.pem -nokeys -caname vally-ca -name vally -caname vally-ca`
password: `crypto`
info: `openssl pkcs12 -info -in trusted.p12`

The files I created can be used for unit tests in poco if desired

File: Crypto/testsuite/src/CryptoTestSuite.cpp [hunk 2]
```diff
@@ -22,6 +23,7 @@ CppUnit::Test* CryptoTestSuite::suite()
 
 	pSuite->addTest(CryptoTest::suite());
 	pSuite->addTest(RSATest::suite());
+	pSuite->addTest(ECTest::suite());
 	pSuite->addTest(DigestEngineTest::suite());
```
[the remark this answers is not in the document]
Comment (aleks-f) [reply Crypto/testsuite/src/CryptoTestSuite.cpp]: yes, of course, that will be done this week

File: Crypto/include/Poco/Crypto/PKCS12Container.h [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	PKCS12Container(const PKCS12Container& cert);
+		/// Copy constructor.
+
+	PKCS12Container& operator = (const PKCS12Container& cert);
+		/// Assignment operator.
+
+	~PKCS12Container();
+		/// Destroys the PKCS12Container.
+
+	EVPPKey getKey() const
```
[the remark this answers is not in the document]
Comment (aleks-f) [reply Crypto/include/Poco/Crypto/PKCS12Container.h]: currently, there's no direct way other than [constructing](https://github.com/pocoproject/poco/blob/crypto/pkcs12/Crypto/include/Poco/Crypto/ECKey.h#L49) eg. ECKey (which is a KeyPair descendant)from PKCS12Container, but it probably makes sense to have KeyPair(const EVPPKey&) constructor

[the remark this answers is not in the document]
Comment (aleks-f) [reply Crypto/include/Poco/Crypto/PKCS12Container.h]: ah, no, scratch that, Keypair is abstract. will look into how to best do this, probably ECKey and RSAKey should have constructors taking `const EVPPKey&`.

**Comment by aleks-f** [comment]
@bachp 1.7.9 is being released today. I plan to have the PKCS12 code complete and backported by the end of this week. 1.8.0 will be released within the next few weeks with PKCS12, hopefully that's ok with you?
/cc @obiltschnig

File: Crypto/testsuite/src/CryptoTestSuite.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (aleks-f) [reply Crypto/testsuite/src/CryptoTestSuite.cpp]: [tests](https://github.com/pocoproject/poco/blob/crypto/pkcs12/Crypto/testsuite/src/PKCS12ContainerTest.cpp) are in, pass via fopen(), but  I still have to check the ifstream version. I'm also not 100% sure about [this](https://github.com/pocoproject/poco/blob/crypto/pkcs12/Crypto/src/PKCS12Container.cpp#L123) yet

File: Crypto/include/Poco/Crypto/PKCS12Container.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (aleks-f) [reply Crypto/include/Poco/Crypto/PKCS12Container.h]: added constructor to [ECKey](https://github.com/pocoproject/poco/blob/crypto/pkcs12/Crypto/include/Poco/Crypto/ECKey.h#L46) and RSAKey, taking `const EVPPKey&`, will add tests too

File: Crypto/src/PKCS12Container.cpp (hunk 1 above)
Comment (aleks-f) [reply Crypto/src/PKCS12Container.cpp]: these were helpful, thanks

**Comment by bachp** [comment]
@aleks-f Having the change in Poco 1.8.0 is fine for me. 👍

**Pull request merged**